compiler/rustc_codegen_llvm/src/intrinsic.rs

```diff
@@ -192,10 +192,23 @@ impl<'ll, 'tcx> IntrinsicCallMethods<'tcx> for Builder<'_, 'll, 'tcx> {
             }
             sym::is_val_statically_known => {
                 let intrinsic_type = args[0].layout.immediate_llvm_type(self.cx);
-                match self.type_kind(intrinsic_type) {
-                    TypeKind::Pointer | TypeKind::Integer | TypeKind::Float | TypeKind::Double => {
+                let kind = self.type_kind(intrinsic_type);
+                match kind {
+                    TypeKind::Pointer
+                    | TypeKind::Integer
+                    | TypeKind::Half
+                    | TypeKind::Float
+                    | TypeKind::Double
+                    | TypeKind::FP128 => {
+                        let type_str = match kind {
```

Comment:
This double match is really awkward. I'd honestly rather the code duplication, or maybe make the `_ =>` branch into a `return` so you can uplift that `let type_str` out of the outer match. Also worth leaving a note why some of these are special cased.